crates/editor/src/editor.rs

```diff
@@ -10119,24 +10179,34 @@ impl Editor {
                 .zip(diagnostic_group.into_iter().map(|entry| entry.diagnostic))
                 .collect();
 
-            Some(ActiveDiagnosticGroup {
-                primary_range,
-                primary_message,
+            self.active_diagnostics.insert(
                 group_id,
-                blocks,
-                is_valid: true,
-            })
-        });
-        self.active_diagnostics.is_some()
+                ActiveDiagnosticGroup {
+                    primary_range,
+                    primary_message,
+                    blocks,
+                },
+            );
+        })
     }
 
-    fn dismiss_diagnostics(&mut self, cx: &mut ViewContext<Self>) {
-        if let Some(active_diagnostic_group) = self.active_diagnostics.take() {
-            self.display_map.update(cx, |display_map, cx| {
-                display_map.remove_blocks(active_diagnostic_group.blocks.into_keys().collect(), cx);
-            });
-            cx.notify();
+    fn focus_diagnostic(&mut self, group_id: usize, cx: &mut ViewContext<Self>) -> bool {
+        if let Some(current_group_id) = self.focused_diagnostic.as_ref() {
+            if group_id == *current_group_id {
+                return false;
+            }
+        }
+        self.focused_diagnostic = Some(group_id);
+        if !self.show_diagnostics_inline {
+            self.refresh_active_diagnostics(cx);
         }
+        true
+    }
+
+    fn unfocus_diagnostics(&mut self, cx: &mut ViewContext<Self>) {
+        self.focused_diagnostic = None;
+        self.refresh_active_diagnostics(cx);
```

Comment:
I find it quite suspicious that we have to do something relatively complex (as iterating over all diagnostics and updating them) when they go in/out of focus.

Curious, why do we need that? 
Could we avoid any diagnostics update and use the old data instead?

Reply:
Whoops. That's totally me making another "See, I'm totally not a UI dev" mistake. The focus/unfocus is a bad name I picked for changing which of the diagnostics `F8` is currently pointing at. I tried to keep it so that if you were searching around F8 style and then toggled the F9 version on and off, the F8 progress/state wouldn't be lost. Oh and also so that F8 can jump to next diagnostic even when F9 is on, which I think is where focus came from (i.e., scroll next one into view). In hindsight, I could see how a real UI person would immediately think window focus/unfocus instead.

As to the specific question on optimizations, my vague memory is "I could optimize this by duplicating some of the logic in refresh_active_diagnostics, or I could just take the (possible, unmeasured, maybe zero) performance hit in completely re-running the refresh logic." Also, this unfocus is only ever called when users dismiss the active F8 style diagnostic and we're already running refresh_active_diagnostics on every key stroke, so :shrug: on that performance hit.